### PR TITLE
feat(egh): add access state to post form

### PR DESCRIPTION
**Pull request by zacjones93**
Access will default to `free` for posts but can now be set to `pro`. On save we update the relevant egghead lesson objects `is_pro_content` and `free_forever` attributes. 

image

gif

### CODE REVIEW

File: apps/egghead/src/app/(content)/posts/_components/edit-post-form-metadata.tsx [hunk 1]
```diff
@@ -47,6 +47,7 @@ import { useSocket } from '@coursebuilder/ui/hooks/use-socket'
 import { MetadataFieldState } from '@coursebuilder/ui/resources-crud/metadata-fields/metadata-field-state'
 import { MetadataFieldVisibility } from '@coursebuilder/ui/resources-crud/metadata-fields/metadata-field-visibility'
 
+import { MetadataFieldAccess } from './metadata-field-access'
```
Comment (zacjones93) [inline apps/egghead/src/app/(content)/posts/_components/edit-post-form-metadata.tsx:50]: This field is very specific to egghead so kept it in the local `_components` folder as apposed to `@coursebuilder/ui/resources-crud/metadata-fields` with the rest of the Metadata fields

File: apps/egghead/src/app/(content)/posts/_components/metadata-field-access.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+} from '@coursebuilder/ui/primitives/select'
+
+export function MetadataFieldAccess({ form }: { form: UseFormReturn<any> }) {
+	return (
+		<FormField
+			control={form.control}
+			name="fields.access"
+			render={({ field }) => (
+				<FormItem className="px-5">
+					<FormLabel>Access</FormLabel>
```
Comment (zacjones93) [inline apps/egghead/src/app/(content)/posts/_components/metadata-field-access.tsx:29]: 'Access' felt like a better name than 'Pro Visibility'

File: apps/egghead/src/lib/egghead.ts [hunk 3]
```diff
@@ -91,6 +91,8 @@ export async function createEggheadLesson(input: {
 		'updated_at',
 		'visibility_state',
 		'guid',
+		'is_pro_content',
+		'free_forever',
```
Comment (zacjones93) [inline apps/egghead/src/lib/egghead.ts:95]: set the default to a free resource as that's the current behavior with posts

File: apps/egghead/src/lib/egghead.ts [hunk 4]
```diff
@@ -161,7 +167,9 @@ export async function updateEggheadLesson(input: {
 			title = $5,
 			slug = $6,
 			guid = $7,
-			summary = $8
+			summary = $8,
+      is_pro_content = $10,
+      free_forever = NOT $10
```
Comment (zacjones93) [inline apps/egghead/src/lib/egghead.ts:172]: `is_pro_content` and `free_forever` are opposites of eachother

**Comment by joelhooks** [comment]
let's set to pro by default

File: apps/egghead/src/app/(content)/posts/_components/edit-post-form-metadata.tsx (hunk 1 above)
Comment (joelhooks) [reply apps/egghead/src/app/(content)/posts/_components/edit-post-form-metadata.tsx:50]: ya, egghead has special needs so I moved it all in. Ultimately it'd be cool if we identified how its different and made changes to the overall composabilty

File: apps/egghead/src/app/(content)/posts/_components/edit-post-form.tsx [hunk 5]
```diff
@@ -61,6 +62,7 @@ export function EditPostForm({
 				postType: post.fields?.postType || 'lesson',
 				body: post.fields?.body,
 				visibility: post.fields?.visibility || 'unlisted',
+				access: post.fields?.access || 'free',
```
Comment (joelhooks) [inline apps/egghead/src/app/(content)/posts/_components/edit-post-form.tsx:65]:
```suggestion
				access: post.fields?.access || 'pro',
```

File: apps/egghead/src/app/(content)/posts/_components/metadata-field-access.tsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+									<SelectItem key={value} value={value}>
+										{value}
+									</SelectItem>
+								)
+							})}
+						</SelectContent>
+					</Select>
+					<FormDescription>
+						Choose whether this post is visible to paying members only or
+						everyone. Default to free.
```
Comment (joelhooks) [inline apps/egghead/src/app/(content)/posts/_components/metadata-field-access.tsx:49]:
```suggestion
						everyone. Default to pro (members only).
```

File: apps/egghead/src/lib/egghead.ts (hunk 4 above)
Comment (joelhooks) [reply apps/egghead/src/lib/egghead.ts:172]: I don't mind it, but it is a little confusing. Likely won't matter in practice.

File: apps/egghead/src/lib/posts-query.ts [hunk 7]
```diff
@@ -392,6 +394,7 @@ export async function writeNewPostToDatabase(input: {
 				title,
 				state: 'draft',
 				visibility: 'unlisted',
+				access: 'free',
```
Comment (joelhooks) [inline apps/egghead/src/lib/posts-query.ts:397]:
```suggestion
				access: 'pro',
```

File: apps/egghead/src/lib/posts.ts [hunk 8]
```diff
@@ -52,6 +56,7 @@ export const PostSchema = ContentResourceSchema.merge(
 			body: z.string().nullable().optional(),
 			state: PostStateSchema.default('draft'),
 			visibility: PostVisibilitySchema.default('public'),
+			access: PostAccessSchema.default('free'),
```
Comment (joelhooks) [inline apps/egghead/src/lib/posts.ts:59]:
```suggestion
			access: PostAccessSchema.default('pro'),
```

File: apps/egghead/src/lib/posts.ts [hunk 9]
```diff
@@ -81,6 +86,7 @@ export const PostUpdateSchema = z.object({
 		body: z.string().optional().nullable(),
 		description: z.string().optional().nullable(),
 		visibility: PostVisibilitySchema.optional().default('public'),
+		access: PostAccessSchema.optional().default('free'),
```
Comment (joelhooks) [inline apps/egghead/src/lib/posts.ts:89]:
```suggestion
		access: PostAccessSchema.optional().default('pro'),
```

**Review by joelhooks — APPROVED** [review]

**Pull request merged**